framework/src/araui/core/ui_core.c

```diff
@@ -0,0 +1,835 @@
+
+	if (!widget) {
+		UI_LOGE("Error: widget is null!\n");
+		return UI_INVALID_PARAM;
+	}
+
+	if (widget->visible) {
+		if (widget->draw_cb) {
+			new_view_port = ui_rect_intersect(draw_area, widget->crop_rect);
+			ui_dal_set_viewport(new_view_port.x, new_view_port.y, new_view_port.width, new_view_port.height);
+			widget->draw_cb((ui_widget_t)widget, dt);
+			ui_dal_set_viewport(draw_area.x, draw_area.y, draw_area.width, draw_area.height);
+		}
+
+		vec_foreach(&widget->children, child, iter) {
+			_ui_draw_widget_recur(child, draw_area, dt);
+		}
+	}
+
+	return UI_OK;
+}
+
+static void _ui_call_tween_finished_cb(void *userdata)
+{
+	ui_widget_body_t *body;
+
+	if (!userdata) {
+		UI_LOGE("error: Invalid Parameter!\n");
+		return;
+	}
+
+	body = (ui_widget_body_t *)userdata;
+	if (body && body->tween_info.tween_finished_cb) {
+		body->tween_info.tween_finished_cb((ui_widget_t)body);
+	}
+}
+
+static void _ui_redraw(uint32_t dt)
+{
+	ui_rect_t *redraw_area;
+	ui_view_body_t *view;
+	int iter;
+	static bool clear_flag = false;
+
+	vec_foreach(ui_view_get_update_list(), redraw_area, iter) {
+		view = ui_view_get_current();
+		if (view) {
+			_ui_draw_widget_recur(view->root, *redraw_area, dt);
+		}
+
+		if (_ui_core_is_visible_global_widget()) {
+			_ui_draw_widget_recur(g_global_widget_info[g_core.visible_event_type], *redraw_area, dt);
+		}
+
+		if (view || _ui_core_is_visible_global_widget()) {
+			ui_dal_redraw(redraw_area->x, redraw_area->y, redraw_area->width, redraw_area->height);
+		}
+
+		clear_flag = true;
+	}
+
+	if (clear_flag) {
+		clear_flag = false;
+		ui_view_update_list_clear();
+	}
+}
+
+static void *_ui_core_thread_loop(void *param)
+{
+	ui_widget_body_t *root;
+	ui_view_body_t *view;
+	static ui_view_body_t *prev_view;
+	struct timespec before;
+	struct timespec now;
+	uint32_t dt;
+
+#if (CONFIG_UI_MAXIMUM_FPS > 0)
+	uint32_t fps_dt = 0;
+	const uint32_t ms_per_frame = 1000 / CONFIG_UI_MAXIMUM_FPS;
+#endif
+
+	memset(&before, 0, sizeof(struct timespec));
+	memset(&now, 0, sizeof(struct timespec));
+
+	clock_gettime(CLOCK_MONOTONIC, &before);
+
+	while (g_core.state == UI_CORE_STATE_RUNNING) {
+		ui_dal_clear();
+
+		clock_gettime(CLOCK_MONOTONIC, &now);
+
+		dt = ((now.tv_sec - before.tv_sec) * 1000) + ((now.tv_nsec - before.tv_nsec) / 1000000);
+		before = now;
+
+#if (CONFIG_UI_MAXIMUM_FPS > 0)
+		fps_dt += dt;
+		if (fps_dt < ms_per_frame) {
+			usleep((ms_per_frame - fps_dt) * 1000);
+		}
+		fps_dt -= ms_per_frame;
+#endif
+
+		view = ui_view_get_current();
+		if (view) {
+			prev_view = view;
+			root = view->root;
+
+			if (_ui_process_widget_recur(root, dt) != UI_OK) {
+				UI_LOGE("error: processing widget recursively failed!\n");
+			}
+
+			ui_render_widget(root);
+		}
+
+		if (_ui_core_is_visible_global_widget()) {
+			_ui_process_widget_recur(g_global_widget_info[g_core.visible_event_type], dt);
+		}
+
+		_ui_redraw(dt);
+
+#if defined(CONFIG_UI_ENABLE_TOUCH)
+		_ui_core_dispatch_touch_event();
```

Comment:
As we discussed before,(even if you didn't agree with me), dispatch touch event is not ui framework's role.
Some of module can display with only specific graphic library(for example, camera).
In this time, below will be enough.
1. Separate dispatcher from uifw
2. apply callback mechanism from dispatcher.

Reply:
No problem. AraUI only depends on the DAL(driver abstract layer).
If DAL is implemented by TizenRT's driver, there is no problem.